File: src/watch.zig
const std = @import("std");
const Io = std.Io;
const report = @import("report.zig");
const cfgmod = @import("config.zig");
const buildmod = @import("build.zig");
const mirrormod = @import("mirror.zig");
const Ctx = report.Ctx;
const Config = cfgmod.Config;

const POLL_MS = 200;

pub fn watch(ctx: *Ctx, initial: Config) !void {
    report.info(ctx, try std.fmt.allocPrint(ctx.arena, "watching {s} {s}", .{
        report.bold(ctx, initial.src),
        report.dim(ctx, "(ctrl-c to stop)"),
    }));

    var scratch = std.heap.ArenaAllocator.init(ctx.gpa);
    defer scratch.deinit();

    var prev_acc: u128 = 0;
    var prev_count: usize = 0;
    var prev_config: i128 = 0;
    var first = true;

    while (true) {
        _ = scratch.reset(.retain_capacity);
        ctx.arena = scratch.allocator();

        const cfg = cfgmod.loadConfig(ctx) catch {
            Io.sleep(ctx.io, Io.Duration.fromMilliseconds(POLL_MS), .awake) catch {};
            continue;
        };

        var acc: u128 = 0;
        var count: usize = 0;
        scanDir(ctx, cfg.src, &acc, &count) catch {};
        const config_mtime = statMtime(ctx, cfgmod.CONFIG_FILE);
        const base_mtime = statMtime(ctx, cfg.base);
        acc +%= mix(cfgmod.CONFIG_FILE, config_mtime);
        acc +%= mix(cfg.base, base_mtime);

        const changed = first or acc != prev_acc or count != prev_count;
        const config_changed = !first and config_mtime != prev_config;

        if (changed) {
            if (config_changed) report.info(ctx, try std.fmt.allocPrint(ctx.arena, "reloaded {s}", .{cfgmod.CONFIG_FILE}));

            if (buildmod.build(ctx, cfg)) |outcome| {
                report.build(ctx, outcome.module_count, outcome.mirror.written, outcome.mirror.removed, outcome.project_changed, outcome.warnings, cfg.out, cfg.mirror.dir, report.clock(ctx));

                if (outcome.mirror.removed != 0 and cfg.mirror.prune_dirs) {
                    Io.sleep(ctx.io, Io.Duration.fromMilliseconds(@intCast(cfg.mirror.prune_delay_ms)), .awake) catch {};
                    const removed = mirrormod.pruneMirror(ctx, cfg) catch 0;
                    if (removed != 0) report.info(ctx, try std.fmt.allocPrint(ctx.arena, "{s} pruned {d} empty folder{s}", .{
                        report.dim(ctx, report.clock(ctx)),
                        removed,
                        if (removed == 1) "" else "s",
                    }));
                }
            } else |e| {
                if (e != error.Reported) report.err(ctx, @errorName(e));
            }
        }

        prev_acc = acc;
        prev_count = count;
        prev_config = config_mtime;
        first = false;

        Io.sleep(ctx.io, Io.Duration.fromMilliseconds(POLL_MS), .awake) catch {};
    }
}

fn scanDir(ctx: *Ctx, dir: []const u8, acc: *u128, count: *usize) !void {
    const a = ctx.arena;
    var handle = ctx.root_dir.openDir(ctx.io, dir, .{ .iterate = true }) catch return;
    defer handle.close(ctx.io);

    var it = handle.iterate();
    while (try it.next(ctx.io)) |entry| {
        const child = try std.fs.path.join(a, &.{ dir, entry.name });
        if (entry.kind == .directory) {
            try scanDir(ctx, child, acc, count);
        } else {
            acc.* +%= mix(child, statMtime(ctx, child));
            count.* += 1;
        }
    }
}

fn statMtime(ctx: *Ctx, path: []const u8) i128 {
    const stat = ctx.root_dir.statFile(ctx.io, path, .{}) catch return 0;
    return stat.mtime.nanoseconds;
}

fn mix(name: []const u8, mtime: i128) u128 {
    const h: u128 = std.hash.Wyhash.hash(0, name);
    return (h *% 1000003) ^ @as(u128, @bitCast(mtime));
}

File: src/watch.ts
import chokidar from "chokidar";
import path from "node:path";

import { build } from "./build.js";
import { CONFIG_FILE, type Config, loadConfig } from "./config.js";
import { pruneMirror } from "./mirror.js";
import * as report from "./report.js";

const DEBOUNCE_MS = 150;

export async function watch(root: string, initial: Config): Promise<void> {
  let config = initial;

  const src = path.join(root, config.src);
  const mirror = path.resolve(root, config.mirror.dir);
  const watcher = chokidar.watch(
    [src, path.join(root, CONFIG_FILE), path.join(root, config.base)],
    {
      ignoreInitial: true,
      ignored: (candidate) => {
        const resolved = path.resolve(candidate);
        return resolved === path.resolve(root, config.out) || resolved.startsWith(mirror);
      },
    },
  );

  let timer: NodeJS.Timeout | undefined;
  let pruneTimer: NodeJS.Timeout | undefined;
  let pendingConfigReload = false;
  let building = false;
  let queued = false;

  // Deferred so Rojo has processed the file removals first: deleting a
  // directory alongside its files crashes Rojo 7.7.0.
  const schedulePrune = () => {
    if (!config.mirror.pruneDirs) return;
    clearTimeout(pruneTimer);
    pruneTimer = setTimeout(() => {
      void pruneMirror(root, config)
        .then((removed) => {
          if (removed) report.info(`${report.dim(report.clock())} pruned ${removed} empty folder${removed === 1 ? "" : "s"}`);
        })
        .catch((error: unknown) => report.error((error as Error).message));
    }, config.mirror.pruneDelayMs);
  };

  const rebuild = async () => {
    if (building) {
      queued = true;
      return;
    }
    building = true;

    try {
      if (pendingConfigReload) {
        pendingConfigReload = false;
        config = await loadConfig(root);
        report.info(`reloaded ${CONFIG_FILE}`);
      }
      const outcome = await build(root, config);
      report.build(outcome, config, report.clock());
      if (outcome.mirror.removed) schedulePrune();
    } catch (error) {
      report.error((error as Error).message);
    } finally {
      building = false;
      if (queued) {
        queued = false;
        void rebuild();
      }
    }
  };

  const schedule = (changed: string) => {
    if (path.basename(changed) === CONFIG_FILE) pendingConfigReload = true;
    clearTimeout(timer);
    timer = setTimeout(() => void rebuild(), DEBOUNCE_MS);
  };

  watcher.on("all", (_event, changed) => schedule(changed));
  watcher.on("error", (error) => report.error(`watch error: ${(error as Error).message}`));

  report.info(`watching ${report.bold(config.src)} ${report.dim("(ctrl-c to stop)")}`);
  await rebuild();

  await new Promise<void>((resolve) => {
    const stop = () => {
      void watcher.close().then(resolve);
    };
    process.once("SIGINT", stop);
    process.once("SIGTERM", stop);
  });
}
